xla/tests/collective_ops_e2e_test.cc

```diff
@@ -1340,6 +1341,290 @@ TEST_F(CollectiveOpsTestE2E, NoAsyncCollectives) {
+    const int64_t num_replicas = 1;
+    if (test_runner().device_count() < num_replicas * num_partitions) {
+      GTEST_SKIP() << "Test requires at least " << num_replicas * num_partitions
+                   << " devices (" << test_runner().device_count()
+                   << " available)";
+    }
+
+    TF_ASSERT_OK_AND_ASSIGN(std::vector<Literal> ref_results,
+                            ExecuteReplicated(hlo_text));
+    ASSERT_EQ(ref_results.size(), 1);
+
+    TF_ASSERT_OK_AND_ASSIGN(
+        std::vector<Literal> results,
+        ExecuteSharded(hlo_text, num_partitions, num_replicas));
+    ASSERT_EQ(results.size(), num_partitions);
+
+    // Slice the unsharded reference results and compare to the sharded case.
+    HloModuleConfig config =
+        GetModuleConfigForTest(/*replica_count=*/num_replicas);
+    auto opts = GetDebugOptionsForTest();
+    opts.set_xla_gpu_enable_triton_gemm(false);
+    config.set_debug_options(opts);
+    config.set_num_partitions(num_partitions);
+    TF_ASSERT_OK_AND_ASSIGN(auto module,
+                            ParseAndReturnVerifiedModule(hlo_text, config));
+    auto dimensions =
+        module->entry_computation()->root_instruction()->shape().dimensions();
+    std::vector<int64_t> root_dims(dimensions.begin(), dimensions.end());
+    std::vector<int64_t> root_dims_per_shard = root_dims;
+    std::vector<int64_t> root_sharded_dims;
+    {
+      HloSharding root_sharding =
+          module->entry_computation()->root_instruction()->sharding();
+      EvaluateShardedDims(root_dims_per_shard, root_sharded_dims,
+                          root_sharding);
+    }
+    ErrorSpec error_spec{1e-4, 1e-4};
+    if (!root_sharded_dims.empty()) {
+      std::vector<int64_t> lower(root_dims_per_shard.size(), 0);
+      std::vector<int64_t> upper(root_dims_per_shard.begin(),
+                                 root_dims_per_shard.end());
+      for (const Literal& result : results) {
+        Literal ref_results_slice = ref_results[0].Slice(lower, upper);
+        EXPECT_TRUE(
+            LiteralTestUtil::Near(ref_results_slice, result, error_spec));
+        for (int m = root_sharded_dims.size() - 1; m >= 0; --m) {
+          if (upper[root_sharded_dims[m]] < root_dims[root_sharded_dims[m]]) {
+            upper[root_sharded_dims[m]] +=
+                root_dims_per_shard[root_sharded_dims[m]];
+            break;
+          } else {
+            upper[root_sharded_dims[m]] =
+                root_dims_per_shard[root_sharded_dims[m]];
+          }
+        }
+        std::transform(upper.begin(), upper.end(), root_dims_per_shard.begin(),
+                       lower.begin(), std::minus<int64_t>());
+      }
+    } else {
+      EXPECT_TRUE(
+          LiteralTestUtil::Near(ref_results[0], results[0], error_spec));
+    }
+  }
+
+ private:
+  // Execute the replicated case.
+  absl::StatusOr<std::vector<Literal>> ExecuteReplicated(
```

Comment:
You are executing the unsharded case here, right? This is not replicated, is it? num_replicas=1 below. 
Maybe call this function `ExecuteUnsharded`